otherlibs/stdlib_alpha/effect.ml

```diff
@@ -896,18 +1295,105 @@ module Make_rec (Ops : Operations_rec)
+       to indicate that the handler is [portable] when it receives
+       [contended] values.
+      
+       We may be able to rewrite this to avoid magic. *)
+      let f h a = f (Obj.magic_portable h) a [@nontail] in
+      let k : (a Modes.Portable.t, b, t, unit) continuation = Obj.magic (fiber f) in
+    (* CR dkalinichenko: here and below, unsafe use of [Obj.magic_portable]:
+       the continuation contains mutable data [Mapping.t].
+       
+       Require [f] to be unique. *)
+      Obj.magic_portable
+        (Continuation k : (a Modes.Portable.t, (b, unit)
+          Result.Portable.Contended.t, unit) Continuation.t)
+
+    let fiber (type a b) f =
+      let f h a = f (Obj.magic_portable h) a [@nontail] in
+      let k : (a Modes.Portable.t, b, t, unit) continuation = Obj.magic (fiber f) in
+      (Continuation k : (a Modes.Portable.t, (b, unit)
+        Result.Contended.t, unit) Continuation.t)
+
+    let portable_fiber_with (type a b es) l f =
+      let f hs a = f (Obj.magic_portable hs) a [@nontail] in
+      let k : (a Modes.Portable.t, b, t, es) continuation = Obj.magic (fiber_with l f) in
+      Obj.magic_portable
+        (Continuation k : (a Modes.Portable.t, (b, es)
+          Result.Portable.Contended.t, es) Continuation.t)
+    
+    let fiber_with (type a b es) l f =
+      let f hs a = f (Obj.magic_portable hs) a [@nontail] in
+      let k : (a Modes.Portable.t, b, t, es) continuation = Obj.magic (fiber_with l f) in
+      (Continuation k : (a Modes.Portable.t, (b, es)
+        Result.Contended.t, es) Continuation.t)
+          
+    let portable_run (type a) f =
+      let f h = f (Obj.magic_portable h) [@nontail] in
+      let res : (a, unit) Result.t = Obj.magic (run f) in
+      Obj.magic_portable (Obj.magic res : (a, unit) Result.Portable.Contended.t)
+    
+    let run (type a) f =
+      let f h = f (Obj.magic_portable h) [@nontail] in
+      let res : (a, unit) Result.t = Obj.magic (run f) in
+      (Obj.magic res : (a, unit) Result.Contended.t)
+      
+    let portable_run_with (type a es) hs f =
+      let f hs = f (Obj.magic_portable hs) [@nontail] in
+      let res : (a, es) Result.t = Obj.magic (run_with hs f) in
+      Obj.magic_portable (Obj.magic res : (a, es) Result.Portable.Contended.t)
+    
+    let run_with (type a es) hs f =
+      let f hs = f (Obj.magic_portable hs) [@nontail] in
+      let res : (a, es) Result.t = Obj.magic (run_with hs f) in
+      (Obj.magic res : (a, es) Result.Contended.t)
+      
+    let perform (type a) (h : t Handler.t) op : a =
+      let op : (a, t) op = Obj.magic op in
+      (* CR dkalinichenko: here and below, we use [Obj.magic_uncontended]
+         to mark a [contended] handler safe to use with [portable]
+         operations. 
+         
+         We may be able to rewrite this to avoid magic. *)
+      perform (op, Obj.magic_uncontended h.h)
+  end
+
+  let portable_fiber (type a b) f =
+    let k : (a, b, t, unit) continuation = Obj.magic (fiber f) in
```

Comment:
This line looks identical to 1373 but has an `Obj.magic`. Is it definitely needed?

Reply:
Removed redundant `Obj.magic` calls.